contracts/proxy/redeem-queue/contract.rs

```
@@ -54,42 +60,211 @@ fn vault_reserve_balance(deps: Deps, hub: &str, vault: &str) -> Result<u128, Err
-    _info: MessageInfo,
-    _vault: String,
-    _index: u64,
+    deps: DepsMut,
+    info: MessageInfo,
+    vault: String,
+    index: u64,
 ) -> Result<Response, Error> {
-    todo!()
+    deps.api.addr_validate(&vault)?;
+
+    let mut queue = RedemptionQueue::new(deps.storage, &vault);
+
+    match queue.get_entry(index) {
+        Some(entry) if entry.address == info.sender => {
+            let (_, amount) = queue.remove_entry(index)?;
+
+            let hub = deps.storage.hub();
+            let synthetic = synthetic_for_vault(deps.as_ref(), &hub, &vault)?;
+
+            let msg = SubMsg::new(cosmwasm_std::BankMsg::Send {
+                to_address: info.sender.to_string(),
+                amount: vec![cosmwasm_std::coin(amount.u128(), synthetic)],
+            });
+
+            Ok(Response::default()
+                .add_attribute("kind", "cancel_entry")
+                .add_attribute("address", &info.sender)
+                .add_attribute("vault", &vault)
+                .add_attribute("index", index.to_string())
+                .add_attribute("amount", amount)
+                .add_submessage(msg))
+        }
+        Some(_) => {
+            bail!("Entry {} does not belong to {}", index, info.sender);
+        }
+        None => {
+            bail!("Entry {} not found", index);
+        }
+    }
 }
 
 pub fn handle_cancel_all(
-    _deps: DepsMut,
-    _info: MessageInfo,
-    _vault: String,
+    deps: DepsMut,
+    info: MessageInfo,
+    vault: String,
 ) -> Result<Response, Error> {
-    todo!()
+    deps.api.addr_validate(&vault)?;
+
+    let mut queue = RedemptionQueue::new(deps.storage, &vault);
+    let cancelled = queue.cancel_user_entries(info.sender.as_ref())?;
+
+    if cancelled.is_empty() {
+        return Ok(Response::default()
+            .add_attribute("kind", "cancel_all")
+            .add_attribute("address", &info.sender)
+            .add_attribute("vault", &vault)
+            .add_attribute("count", "0"));
+    }
```

I think we should return an error in this case.

If someone issued this message expecting to cancel one or more entries, cancelling no entries would be unexpected and should be handled as an error.

---

I could go either way here. I thought of it like "I want to cancel all my entries regardless of how many I have" but it's no problem to change it.